bazaarvoice/emodb
# Service Management and Failure Handling

## Github Issue #

`None`

## What Are We Doing Here?

We are actually doing a couple things here, which I know is generally against team policy. However, most of these changes are hard to separate and depend on each other. 

1. I implemented `ResilientService`, which is Guava service that will restart itself if it dies. This is most useful for services that run KStreams applications, which are not highly available on their own.
2. I I replatformed `MegabusRefResolver` and `MissingRefDelayProcessor` to be resilient services, which allows them to reboot themselves 30 seconds after failure. Additionally, their common configuration and setup was centralized in `KafkaStreamsApplication`.
3. I significantly refactored the boot workflow to be more robust and make boot failure nearly impossible.
4. I refactored the boot process of `MegabusRefProducer` to subscribe before boot, but not poll until after the boot is complete.
 
## How to Test and Verify

The actual observable changeset of this is small, as failures do not normally happen. The best way to test `ResilientService` is to modify the code in `MegabusRefResolver` to intentionally throw exceptions, then observe it restarting after an exception. You can confirm that the `MegabusRefProducer` is not running by running `curl -XPOST localhost:8083/tasks/leader` and ensuring that there has not been a leader elected for the `MegabusRefProducer`.

## Risk

### Level 

`Medium`

### Required Testing

`Regression` and `Manual`

### Risk Summary

The main risk here is that we broke something during this refactor that wasn't broken before. I'm confident we will catch any potential issues during our stash comparison testing though. 

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
